runtime/array.c

```diff
@@ -308,12 +308,14 @@ CAMLprim value caml_make_array(value init)
 
 /* [wo_memmove] copies [nvals] values from [src] to [dst]. If there is a single
    domain running, then we use [memmove]. Otherwise, we copy one word at a
-   time. Since the [memmove] implementation does not guarantee that the writes
-   are always word-sized, we explicitly perform word-sized writes to avoid
-   mixed-mode accesses.
-
-   [MM]: TODO enforcing the memory model also requires fences. See note in
-   memory.c */
+   time.
+
+   Since the [memmove] implementation does not guarantee that the writes are
+   always word-sized, we explicitly perform word-sized writes of the release
+   kind to avoid mixed-mode accesses. Performing release writes should be
+   sufficient to prevent smart compilers from coalesing the writes into vector
+   writes, and hence prevent mixed-mode accesses. [MM].
```

Comment:
According to the comments in memory.c, you might also need a memory barrier.  Presumably one barrier before the whole loop?

Reply:
That's exactly what I was thinking over dinner. Adding a barrier now.  

As an aside, I'm adding `[MM]` in comments to those places in code with changes relevant to the memory model. These will need to be sanity checked for the arm64 backend.